No-seek implementation of HttpResponse Image.open

mfitzp opened the pull request:
This is an alternative implementation for `Image.open()`
to support opening of images from HttpResponse objects. It
uses `.seekable()` to determine whether the file supports
seeking (with failover for this attribute not being supported
e.g. in `urllib2` response objects).

The prefix is chopped off as normal and pre-pended back onto
the bytes fed into the `io.IOBytes` object. This avoids loading
the image object totally (duplicating it in memory) until
we are sure we have a factory for the file.

mfitzp commented:
This is an improvement on the previously merged PR https://github.com/python-pillow/Pillow/pull/1151

homm commented on PIL/Image.py:
```diff
@@ -2250,9 +2250,9 @@ def open(fp, mode="r"):
-    except (AttributeError, io.UnsupportedOperation):
-        fp = io.BytesIO(fp.read())
+        requires_iobytes_wrapper = not fp.seekable()
+    except:
```
Why not leave AttributeError?

homm commented on PIL/Image.py:
```diff
@@ -2250,9 +2250,9 @@ def open(fp, mode="r"):
         filename = ""
 
     try:
-        fp.seek(0)
```
This changes behavior. Previously all files was forcedly sought to the 0.

aclark4life commented:
Also the Travis CI build failed on all versions. Worst case we can leave in the existing PR and iterate on this feature during the 2.9.0 development cycle.

mfitzp commented:
@homm The missing `AttributeError` was an oversight. The change in behaviour to not forcibly `fp.seek(0)` on every file was in response to comments on https://github.com/python-pillow/Pillow/pull/1151 complaining about changes to the pre-existing behaviour (before that PR was merged).

Previously there was no seek to the start of the file before reading the prefix. I'm not sure if that makes sense - since there is a `fp.seek(0)` before loading - but it's how it was.

mfitzp commented:
The latest PR adds the `AttributeError` and addresses the source of the earlier test failure (namely wrapping the HttpResponse object multiple times, resulting in an empty file). The code is a bit less clear than the `fp.seek(0)` `try..except` approach but addresses the issues raised regarding that:

- do not change existing behaviour by seeking before reading the prefix
- do not read the entire file in until we're (semi-)confident we can read it (using the prefix) 

However, while the earlier error is fixed Travis is now is now showing failures on Python 2.x with `ERROR: TestFileLibTiff.test_fp_leak` which is due (it turns out) to `.seekable()` not being a valid property on Python2.7 file objects (they don't appear to inherit from `io.IOBase`).

So there are two options:
- use the `fp.seek(0) try..except` but postpone the wrapping (changes the pre-existing behaviour, adds the extra syscall for the first seek, but avoids the memory usage of opening non-valid files)
- implement a more complicated set of tests - test first for `.seekable` then if that fails test for the existence of `.seek` attribute (no call) ...the first will catch io.IOBase objects so the existence of ineffective `.seek()` on them won't matter.
- forget this entirely and go with the earlier PR using `fp.seek(0) try..except`

I'm completely on the fence!

aclark4life commented:
@mfitzp Sounds like we are waiting on this then and I'll assume you are okay with the previous merge shipping as-is (vs removing it, which I'd prefer not to do at this point.)

homm commented on PIL/Image.py:
```diff
@@ -2250,9 +2250,11 @@ def open(fp, mode="r"):
+    except AttributeError:
+        # Not a subclass of io.IOBase; probably Python 2.7 file (or urllib object)
+        # Check we have a seek fn on the object
+        requires_iobytes_wrapper = not callable( getattr(fp, "seek", None) )
```
What if there is no `seekable` attribute, but calling `seek` will raise `io.UnsupportedOperation`?

homm commented:
@mfitzp you are right about there was no `seek(0)` formerly.  But now I also see the `fp.seek(0)` before the loading and it looks to me what it is better to seek to the start when possible.

mfitzp commented:
@homm Regarding the note: that is a good point although it would require an object that is using `io` Exceptions without inheriting from `io.IOBase`. I'm not sure if that's likely. The check would get even more complicated.

If the consensus is to implement an initial `fp.seek(0)` then that's in the other PR. That leaves the other issue of reading the entire file before we know if we can handle it. One option would be to check as before, but maintain the flag:

    try:
        fp.seek(0)
    except:
        requires_iobytes_wrapper = True
    else:
        requires_iobytes_wrapper = False

Note that this means files and `HttpResponse` objects will be handled slightly differently (files will be zero-seeked before reading the prefix, while `HttpResponse` objects wont). It depends whether the issue of temporarily duplicating a file in memory outweighs that discrepency.

Back on the fence :8ball:

radarhere commented:
Just to let you know, this PR now has merge conflicts.

wiredfool commented:
I can manage merge conflicts at review/merge time.

wiredfool commented:
The recent changes have made it harder to accomplish what this pr is trying to do, since the locations where it's wrapping fp in an io.BytesIO object has been deduped into a function. It's going to make it unpleasant to conditionally read and only do one read of the file. 

Since this has sat for ~6 months, I'm assuming that we don't have any overwhelmingly compelling cases where it's required and that the current version is working well enough.

mfitzp commented:
This PR was an alternative approach to the other merge (on mobile can't
link). I prefer the other method, but this was for backwards compatibility
with reported existing uses. That said I think the reported uses were
relying on a slightly odd behaviour (reading from the current file pos then
subsequent re-seeking).

In short, please feel free to delete!
On 11 Sep 2015 14:34, "wiredfool" <notifications@github.com> wrote:

The recent changes have made it harder to accomplish what this pr is trying
to do, since the locations where it's wrapping fp in an io.BytesIO object
has been deduped into a function. It's going to make it unpleasant to
conditionally read and only do one read of the file.

Since this has sat for ~6 months, I'm assuming that we don't have any
overwhelmingly compelling cases where it's required and that the current
version is working well enough.

—
Reply to this email directly or view it on GitHub
<https://github.com/python-pillow/Pillow/pull/1157#issuecomment-139534039>.

wiredfool commented:
Ok. Thanks for the explanation.
